contracts/extensions/CoinMachine.sol

```diff
@@ -0,0 +1,202 @@
+/*
+  This file is part of The Colony Network.
+
+  The Colony Network is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  The Colony Network is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with The Colony Network. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+pragma solidity 0.5.8;
+pragma experimental ABIEncoderV2;
+
+import "./../../lib/dappsys/math.sol";
+import "./../colony/IColony.sol";
+import "./../common/ERC20Extended.sol";
+
+// ignore-file-swc-108
+
+
+contract CoinMachine is DSMath {
+
+  IColony colony;
+
+  constructor(address _colony) public {
+    colony = IColony(_colony);
+  }
+
+  // Storage
+  address purchaseToken; // The token in which we receive payments, 0x0 for eth
+
+  uint256 periodLength; // Duration of a sale period in seconds (e.g. one hour)
+  uint256 windowSize; // Number of periods in the price averaging window in integers (e.g. 10)
+
+  uint256 targetPerPeriod; // Target number of tokens to sell in a period
+  uint256 maxPerPeriod; // Maximum number of tokens sellable in a period
+
+  uint256 activePeriod; // The active sale period
+  uint256 activePrice; // The active sale price (a WAD-denominated ratio)
+
+  uint256 tokensSold; // Tokens sold in the active period
+  uint256 activeIntake; // Payment received in the active period
+
+  mapping (uint256 => uint256) pastIntakes; // Payment received in past periods
+
+  // Public
+
+  function initialise(
+    address _purchaseToken,
+    uint256 _periodLength,
+    uint256 _windowSize,
+    uint256 _targetPerPeriod,
+    uint256 _maxPerPeriod,
+    uint256 _startingPrice
+  )
+    public
+  {
+    require(activePeriod == 0, "coin-machine-already-initialised");
+
+    require(_periodLength > 0, "coin-machine-period-too-small");
+    require(_windowSize > 0, "coin-machine-window-too-small");
+    require(_windowSize <= 256, "coin-machine-window-too-large");
+    require(_targetPerPeriod > 0, "coin-machine-target-too-small");
+    require(_maxPerPeriod >= _targetPerPeriod, "coin-machine-max-too-small");
+
+    purchaseToken = _purchaseToken;
+
+    periodLength = _periodLength;
+    windowSize = _windowSize;
+
+    targetPerPeriod = _targetPerPeriod;
+    maxPerPeriod = _maxPerPeriod;
+
+    activePrice = _startingPrice;
+    activePeriod = getCurrentPeriod();
+
+    uint256 startingIntake = wmul(targetPerPeriod, _startingPrice);
+    for (uint256 i; i < _windowSize; i++) {
+      pastIntakes[i] = startingIntake;
```

Comment:
I'm still not a huge fan of this. But on the plus side, it does mean that the gas bottleneck is here, and not anywhere else, which is a good situation to be in (if it could be initialised to introduce a bottleneck later on in the flow, it could get stuck).

Reply:
Yup I agree, I think one-off setup costs aren't the end of the world.